### PR TITLE
fix #230: Fix schema (in)sensitive casing issue

### LINKED ISSUE #230: Grate is applying a case-sensitive comparison to Schema name in a case-insensitive SQL Server.
I'm migrating an existing RH db over to grate (on a case-insensitive SQL 2019 server), and when migrated with `grate --schema=roundhouse [other options elided]` it fails with:

```
Unhandled exception: Microsoft.Data.SqlClient.SqlException (0x80131904): There is already an object named 'roundhouse' in the database.
CREATE SCHEMA failed due to previous errors.
   at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at Microsoft.Data.SqlClient.SqlCommand.InternalEndExecuteNonQuery(IAsyncResult asyncResult, Boolean isInternal, String endMethod)
   at Microsoft.Data.SqlClient.SqlCommand.EndExecuteNonQueryInternal(IAsyncResult asyncResult)
   at Microsoft.Data.SqlClient.SqlCommand.EndExecuteNonQueryAsync(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location ---
   at grate.Migration.AnsiSqlDatabase.ExecuteNonQuery(DbConnection conn, String sql, Nullable`1 timeout) in /home/runner/work/grate/grate/grate/Migration/AnsiSqlDatabase.cs:line 664
   at grate.Migration.AnsiSqlDatabase.ExecuteNonQuery(DbConnection conn, String sql, Nullable`1 timeout) in /home/runner/work/grate/grate/grate/Migration/AnsiSqlDatabase.cs:line 664
   at grate.Migration.AnsiSqlDatabase.CreateRunSchema() in /home/runner/work/grate/grate/grate/Migration/AnsiSqlDatabase.cs:line 251
   at grate.Migration.AnsiSqlDatabase.RunSupportTasks() in /home/runner/work/grate/grate/grate/Migration/AnsiSqlDatabase.cs:line 240
   at grate.Migration.GrateMigrator.CreateGrateStructure(IDbMigrator dbMigrator) in /home/runner/work/grate/grate/grate/Migration/GrateMigrator.cs:line 212
   at grate.Migration.GrateMigrator.Migrate() in /home/runner/work/grate/grate/grate/Migration/GrateMigrator.cs:line 85
   at grate.Commands.MigrateCommand.<>c__DisplayClass0_0.<<-ctor>b__0>d.MoveNext() in /home/runner/work/grate/grate/grate/Commands/MigrateCommand.cs:line 48
```

Correcting the script to `grate --schema=RoundhousE [other options elided]` works correctly.

**Desktop (please complete the following information):**
 - OS: Windows 11
 - Version 1.3 and 1.4

### PR DESCRIPTION
**Pull request by wokket**
Had some merge issues but think I'm there.

Assorted changes to DBMS's that are (generally) case insensitive.

PGSql is a little interesting with case-sensitive tables but case insensitive schemas (as currently implemented), but so be it.

There's possibly an option to bypass the `Exists` checks entirely for those DBMS's that support `create table if not exists` syntax, but that's a bridge too far for me time-wise atm.

### CODE REVIEW

File: grate/Migration/AnsiSqlDatabase.cs [hunk 1]
```diff
@@ -256,7 +256,7 @@ private async Task<bool> RunSchemaExists()
     {
         string sql = $"SELECT s.schema_name FROM information_schema.schemata s WHERE s.schema_name = '{SchemaName}'";
         var res = await ExecuteScalarAsync<string>(ActiveConnection, sql);
-        return res == SchemaName;
+        return res != null; // #230: If the server found a record that's good enough for us
```
Comment (erikbra) [inline grate/Migration/AnsiSqlDatabase.cs:259]: Will it be `null`, or can it be `DBNull`?

Comment (wokket) [reply grate/Migration/AnsiSqlDatabase.cs:259]: AFAIK handling the DBNull returned from the sql client is Dapper's job, and it converts to either `null` or `default()` depending on the type it's trying to map to.

If you have any info to the contrary I'd be very keen to know about it!

**Comment by wokket** [comment]
G'day @erikbra , just wondering if there's anything holding up this fix from your side? I tried to convert another existing Roundhouse/SQL server project to grate recently and still have to force v1.3.2 to get it to work...  Hoping this and a few other fixes will get me closer to the line

**Comment by erikbra** [comment]
Hi, @wokket  - thanks for being patient. No, nothing other than myself holding this up (but that is a big enough hurdle by itself sometimes). I resolved some merge conflicts, and I'm waiting for the tests to run, and I'm ready to merge if they run as they should.

It's been hard to prioritise a lot of time on grate the last months, sorry.

**Comment by erikbra** [comment]
Seems like tests are failing on MariaDB and SqlServer, unfortunately, @wokket

**Comment by wokket** [comment]
OK, after fighting with inconsistent test runs locally for a few hours I think we still have some stuff to sort out here.

In the current `main` codebase, we are still applying a case-sensitive check for whether the grate schema exists, regardless of server casing rules, so #230 is still in play :(

The new check for whether a table (like `Version`) exists is now applying a case-_in_sensitive search, regardless of server casing rules, which means the scenario I'm testing here (run a migration against a schema, then again a second time with different casing, replicating what we did to try and upgrade our roundhouse project to grate)  fails because the exists check for `Version` says it exists, but all the following uses of the table fail as it can't find the table as passed-in.

Do  we want to change the table exists checks to return both the schema and table _as found in the database_ and use that for the rest of the migration? 

I understand the case-insensitive lookups as a workaround for bug #245, and keeping back-compat with existing database already migrated is obviously paramount, but it feels like forcing case insensitivity in some places, and respecting the target casing rules are fundamentally incompatible :(

**Comment by cocytus** [comment]
Related: After one year, I don't think i issue #245 is actually fixed for postgres. I tried the latest 1.5.4 release, and exactly as a year ago, grate creates "ScriptsRun[Error]" and "Version" tables, even though "scriptsrun" and "version" exists.

**Comment by erikbra** [comment]
This one should be solved by. other means now :)

**Pull request closed without merging**